src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageSource.java

```diff
@@ -0,0 +1,660 @@
+		}
+		return resourceHolder;
+	}
+
+	protected void createConsumer() {
+		this.consumer = this.consumerFactory.createConsumer(this.groupId, this.clientId, null);
+		synchronized (this.consumerMonitor) {
+			this.consumer.subscribe(Arrays.asList(this.topics), new ConsumerRebalanceListener() {
+
+				@Override
+				public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+					KafkaMessageSource.this.partitions = Collections.emptyList();
+				}
+
+				@Override
+				public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+					KafkaMessageSource.this.partitions = new ArrayList<>(partitions);
+				}
+
+			});
+		}
+	}
+
+	@Override
+	public synchronized void destroy() throws Exception {
+		if (this.consumer != null) {
+			Consumer<K, V> consumer2 = this.consumer;
+			this.consumer = null;
+			synchronized (this.consumerMonitor) {
+				consumer2.close(30, TimeUnit.SECONDS);
+			}
+		}
+	}
+
+	/**
+	 * AcknowledgmentCallbackFactory for KafkaAckInfo.
+	 *
+	 * @param <K> the key type.
+	 * @param <V> the value type.
+	 *
+	 */
+	public static class KafkaAckCallbackFactory<K, V> implements AcknowledgmentCallbackFactory<KafkaAckInfo<K, V>> {
+
+		@Override
+		public AcknowledgmentCallback createCallback(KafkaAckInfo<K, V> info) {
+			return new KafkaAckCallback<K, V>(info);
+		}
+
+	}
+
+	/**
+	 * AcknowledgmentCallback for Kafka.
+	 *
+	 * @param <K> the key type.
+	 * @param <V> the value type.
+	 *
+	 */
+	public static class KafkaAckCallback<K, V> implements AcknowledgmentCallback, Acknowledgment {
+
+		private final Log logger = LogFactory.getLog(getClass());
+
+		private final KafkaAckInfo<K, V> ackInfo;
+
+		private volatile boolean acknowledged;
+
+		private boolean autoAckEnabled = true;
+
+		public KafkaAckCallback(KafkaAckInfo<K, V> ackInfo) {
+			this.ackInfo = ackInfo;
+		}
+
+		@Override
+		public void acknowledge(Status status) {
+			Assert.notNull(status, "'status' cannot be null");
+			if (this.acknowledged) {
+				throw new IllegalStateException("Already acknowledged");
+			}
+			synchronized (this.ackInfo.getConsumerMonitor()) {
+				try {
+					ConsumerRecord<K, V> record = this.ackInfo.getRecord();
+					switch (status) {
+					case ACCEPT:
+					case REJECT:
+						commitIfPossible(record);
+						break;
+					case REQUEUE:
+						rollback(record);
+						break;
+					default:
+						break;
+					}
+				}
+				catch (WakeupException e) {
+					throw new IllegalStateException(e);
+				}
+				finally {
+					this.acknowledged = true;
+					if (!this.ackInfo.isAckDeferred()) {
+						this.ackInfo.getOffsets().get(this.ackInfo.getTopicPartition()).remove(this.ackInfo);
+					}
+				}
+			}
+		}
+
+		private void rollback(ConsumerRecord<K, V> record) {
+			if (this.ackInfo.isMyTransaction()) {
+				this.ackInfo.getProducer().abortTransaction();
+				this.ackInfo.getProducer().close();
+				TransactionSynchronizationManager.unbindResource(this.ackInfo);
```

Comment:
This somehow breaks my idea of transactions. so I need more time to figure out what's going on here and why it is so complicated. I'm not sure why would one use producer transaction here without any transaction aspect around `MS.receive()`, but if that is really the case because Apache Kafka suggest such a model, I'm not sure why would one then have a synchronization and would manage it manually like this code, but don't rely on the outer TX boundaries...